### ISSUE TITLE
Canvas dark mode

### ISSUE THREAD

**Issue by KelvinNovais** [flxzt/rnote#1293]
**Please submit only one feature request in one issue!**  
It would very useful to apply a dark theme to the canvas.

**Is your feature request related to a problem? Please describe.**  
No.

**Describe the solution you'd like**  
This could be achieved by inverting the background color and the strokes colors. Since these functionalities are already implemented, there's just the need to call both of them together.

**Describe alternatives you've considered**  
I think it's interesting to have:

1. a button to manually toggle the dark mode, independent of the user theme;
2. an option on the settings to toggle the dark mode according to the user theme (light or dark)

**Additional context**
As i mentioned previously, this is achieved just by inverting colors of the strokes and the canvas background:

![Light theme](https://github.com/user-attachments/assets/80ea24cf-c7d8-491a-9dbc-a7e46154958f)

![Dark theme](https://github.com/user-attachments/assets/8b78e67f-ade7-4d3e-81c2-0b53188a0972)

**Comment by Kneemund** [comment]
I [considered this](https://github.com/flxzt/rnote/issues/635#issuecomment-1523396001) while working on the color inversion, but ultimately decided against it, because:

- changing the theme of your desktop environment probably shouldn't modify all strokes of your current document;
- users have too much freedom in picking their document colors for this to always turn out decent - e.g. you probably wouldn't want your background to be a dark brown when starting out in a light yellow.

**Comment by KelvinNovais** [comment]
> Adding a button for that could probably work though.

I agree that a button for manual changes is enough, and switching the canvas color mode according to the user theme might not be the best in some cases, it's up to you.

> I [considered this](https://github.com/flxzt/rnote/issues/635#issuecomment-1523396001) ...

Differently from a document you are importing, that has an undefined color palette, I suggest creating two (or more) color palettes for dark and light theme, including the background and the primary colors (those you have on the top menu). In this way, the user can choose what looks better for light/dark mode, possibly avoiding this:

>  e.g. you probably wouldn't want your background to be a dark brown when starting out in a light yellow

Colors that are not included on the color palette can be simply inverted.

This is not straight forward as a simple color inversion, which would work fine for me.

**Comment by ChrisRenfrow** [comment]
I would love this feature as someone who switches system dark/light mode contextually based on my current environment. 

I think a very simple implementation that could go a long way toward satisfying the request would be to allow the user to manually define a dark and light mode canvas color, which would automatically switch based on the selected scheme (*App Menu > Color Scheme*).

A further enhancement on top of this solution could use some kind of on-the-fly (non-destructive) color display adjustment to ensure a minimum contrast ratio is achieved given the currently used palette and background color.

I am open to taking a shot at contributing the first part if that's something the maintainers think is reasonable.

**Comment by KelvinNovais** [comment]
I'd like to point that [Papers](https://flathub.org/apps/org.gnome.Papers) has a "Night Mode", which:

1. forces the application to dark mode and,
2. inverts the document color.

This behavior is my first suggestion, I think it's easier to implement, although the simplest solution.

**Comment by ChrisRenfrow** [comment]
Yes, I think that would be most simple as well. Though, in the example given by @Kneemund, I can understand why a simple inversion would only be a partial solution in the case that users have selected a background color that wouldn't look very nice inverted. Allowing the dark-mode color to be overridden seems like a necessary control to have if this is the route taken.

There may also be other more reasonable color-space calculations to consider when picking a complimentary dark mode color outside of inversion, but this isn't really something I'm deeply familiar with (yet).

At any rate, I'll be getting this project set-up on my machine today so I can find my way around and get started. :)

**Comment by ChrisRenfrow** [comment]
Okay, so I've dug into this a little more tonight and, I am ashamed to admit that I only just now noticed the "Invert Color Brightness" button in the settings, which is much better than manually setting the background and pattern colors as I have been. 🤡 

Now that I have a more complete understanding, here's my more focused suggestion.

I think the "Invert Color Brightness" setting could be reworked from being a button action to a drop-down selection with the following options:
- _Enabled_
- _Disabled_*
- _Auto (based on Color Scheme)_

(* default)

On top of providing room for the feature in question, this seems like a slightly nicer pattern than having a button that fires a one-off action in the settings, as it explicitly communicates whether the inversion is active or not.

Does that seem reasonable to you @KelvinNovais?

I noticed `@Kneemund` is marked as inactive in the maintainers list so I'm pinging @flxzt and/or @Doublonmousse for input/direction whenever either of them have a moment (no rush).

**Comment by ChrisRenfrow** [comment]
> I mentioned this on the issue description

Yep, I just wasn't making the connection because I wasn't aware of the feature(s). :)

> inverting the background color, doesn't invert the strokes color

Right, which is where a button action (as you said would work for you earlier in the thread) would make sense when altering the stroke colors of the entire document as needed.

I think an item in the Canvas Menu called "Invert palette and stroke brightness" would do the trick, selecting it again would effectively undo the operation (in addition to the actual undo operation).

This replaces:
1. Selecting all strokes with the selection tool
2. Selecting "Invert Color Brightness of All Selected Strokes" in selection tool menu
3. Manually adjusting all palette options to match

I think this will work for most if not all cases where this operation would be preferable to the manual steps.

---

And with that, I think I have enough to get started. :) Of course, if the maintainers have any guidance to offer I'll gladly accept it here or in the PR once I get around to making it.

**Comment by felixmueh** [comment]
I just want to add my vote for an automatic system theme based toggle of both the background as well as the palette and strokes. This is actually my chief annoyance with the software. I tend to switch themes quite often when I change environments and it is basically impossible to do a theme change quickly and on the fly with rnote, which otherwise is amazing software.

**Comment by felixmueh** [comment]
A different approach that would also address #998 (inverting image colors) is to simply add a post processing inversion filter that can be set in the settings to on / off / follow systemtheme. 

What would make this more complicated is that this would either have to be applied to every single place that palette color is displayed or it would have to accept inconsistencies in the palette colors and the colors drawn ...
